core/bluetooth/src/main/kotlin/edu/stanford/spezi/core/bluetooth/domain/BLEPairedDevicesStorage.kt

```diff
@@ -0,0 +1,125 @@
+package edu.stanford.spezi.core.bluetooth.domain
+
+import android.bluetooth.BluetoothAdapter
+import android.bluetooth.BluetoothDevice
+import edu.stanford.spezi.core.bluetooth.data.model.BLEDevice
+import edu.stanford.spezi.core.coroutines.di.Dispatching
+import edu.stanford.spezi.core.logging.speziLogger
+import edu.stanford.spezi.core.utils.TimeProvider
+import edu.stanford.spezi.modules.storage.di.Storage
+import edu.stanford.spezi.modules.storage.key.KeyValueStorage
+import edu.stanford.spezi.modules.storage.key.getSerializableList
+import edu.stanford.spezi.modules.storage.key.putSerializable
+import kotlinx.coroutines.CoroutineExceptionHandler
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.launch
+import javax.inject.Inject
+
+@Suppress("MissingPermission")
+internal class BLEPairedDevicesStorage @Inject constructor(
+    private val bluetoothAdapter: BluetoothAdapter,
+    private val bleDevicePairingNotifier: BLEDevicePairingNotifier,
+    @Storage.Encrypted
+    private val encryptedKeyValueStorage: KeyValueStorage,
+    private val timeProvider: TimeProvider,
+    @Dispatching.IO private val ioScope: CoroutineScope,
+) {
+    private val logger by speziLogger()
+    private val coroutineExceptionHandler = CoroutineExceptionHandler { _, error ->
+        logger.e(error) { "Error executing paired devices storage operations" }
+    }
+
+    private val _pairedDevices = MutableStateFlow(emptyList<BLEDevice>())
+    val pairedDevices = _pairedDevices.asStateFlow()
+
+    init {
+        refreshState()
+        observeUnpairingEvents()
+    }
+
+    fun updateDeviceConnection(device: BluetoothDevice, connected: Boolean) = execute {
+        if (isPaired(device).not()) return@execute
+        val currentDevices = getCurrentStoredDevices()
+        currentDevices.removeAll { it.address == device.address }
+        val newDevice = BLEDevice(
+            address = device.address,
+            name = device.name,
+            connected = connected,
+            lastSeenTimeStamp = timeProvider.currentTimeMillis()
+        )
+
+        update(devices = currentDevices + newDevice)
+    }
+
+    private fun refreshState() = execute {
+        val systemBoundDevices = bluetoothAdapter.bondedDevices
```

Comment:
`bluetoothAdapter.bondedDevices` can return null depending on the state of the Bluetooth adapter.

```suggestion
        val systemBoundDevices = bluetoothAdapter.bondedDevices ?: return@execute
```

Reply:
Thanks 🙌 commited your suggestion. But in general execute launches the io scope which already catches the exceptions via the coroutine exception handler, this for instance can also throw in case of missing permission, hence it is safe to use execute

Reply:
Thanks for the explanation, good to know! 👍